storage/range_tree.go

```diff
@@ -20,29 +20,286 @@ package storage
+func (n *RangeTreeNode) set(batch engine.Engine, ms *engine.MVCCStats, timestamp proto.Timestamp) error {
+	if err := engine.MVCCPutProto(batch, ms, engine.RangeTreeNodeKey(n.Key), timestamp, nil, (*proto.RangeTreeNode)(n)); err != nil {
+		return err
+	}
+	return nil
+}
+
+// set saves the RangeTreeRoot to the db.
+func (t *RangeTree) set(batch engine.Engine, ms *engine.MVCCStats, timestamp proto.Timestamp) error {
+	if err := engine.MVCCPutProto(batch, ms, engine.KeyRangeTreeRoot, timestamp, nil, (*proto.RangeTree)(t)); err != nil {
+		return err
 	}
+	return nil
+}
 
-	node := &proto.RangeTreeNode{
+// SetupRangeTree creates a new RangeTree.  This should only be called as part of
+// store.BootstrapRange.
+func SetupRangeTree(batch engine.Engine, ms *engine.MVCCStats, timestamp proto.Timestamp, startKey proto.Key) error {
+	tree := (*RangeTree)(&proto.RangeTree{
+		RootKey: startKey,
+	})
+	node := (*RangeTreeNode)(&proto.RangeTreeNode{
 		Key:   startKey,
 		Black: true,
+	})
+	if err := node.set(batch, ms, timestamp); err != nil {
+		return err
 	}
-
-	if err := engine.MVCCPutProto(batch, ms, engine.RangeTreeNodeKey(startKey), timestamp, nil, node); err != nil {
+	if err := tree.set(batch, ms, timestamp); err != nil {
 		return err
 	}
+	return nil
+}
+
+// GetRangeTree fetches the RangeTree proto.
+func GetRangeTree(batch engine.Engine, timestamp proto.Timestamp) (*RangeTree, error) {
+	tree := &proto.RangeTree{}
+	ok, err := engine.MVCCGetProto(batch, engine.KeyRangeTreeRoot, timestamp, nil, tree)
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		return nil, util.Errorf("Could not find the range tree:%s", engine.KeyRangeTreeRoot)
+	}
+	return (*RangeTree)(tree), nil
+}
 
-	if err := engine.MVCCPutProto(batch, ms, engine.KeyRangeTreeRoot, timestamp, nil, tree); err != nil {
-		return err
+// getRangeTreeNode return the RangeTree node for the given key.  If the key is
+// nil, an empty RangeTreeNode is returned.
+func getRangeTreeNode(batch engine.Engine, timestamp proto.Timestamp, key *proto.Key) (*RangeTreeNode, error) {
+	node := &proto.RangeTreeNode{}
+	if key == nil {
+		return (*RangeTreeNode)(node), nil
 	}
+	ok, err := engine.MVCCGetProto(batch, engine.RangeTreeNodeKey(*key), timestamp, nil, node)
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		return nil, util.Errorf("Could not find the range tree node:%s",
+			engine.RangeTreeNodeKey(*key))
+	}
+	return (*RangeTreeNode)(node), nil
+}
+
+// getRootNode returns the RangeTree node for the root of the RangeTree.
+func (t *RangeTree) getRootNode(batch engine.Engine, timestamp proto.Timestamp) (*RangeTreeNode, error) {
+	return getRangeTreeNode(batch, timestamp, &t.RootKey)
+}
 
+// InsertRange adds a new range to the RangeTree.  This should only be called
+// from operations that create new ranges, such as range.splitTrigger.
+func (t *RangeTree) InsertRange(batch engine.Engine, ms *engine.MVCCStats, timestamp proto.Timestamp, key proto.Key) error {
```

Comment:
I thought the range tree node for a range was going to be stored locally at the range. If that is the case, don't these methods need to be doing distributed operations? The way this code is currently set up it looks like all of the range tree stuff will be stored within a single `engine.Engine`.